README.md
NumSem v0.3.0 (2018.10.15)
======

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1256765.svg)](https://doi.org/10.5281/zenodo.1256765)

An application in C++ for solving diverse problems in numerical semigroups:
* Numerical semigroup membership
* Sylvester denumerant
* Frobenius number
* Set of gaps
* Apéry set (*in development*)
* Minimal system of generators (*in development*)

Building
--------

#### C++

```
make
cd bin
numsem


CHANGELOG.md

# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.3.0 (2018.10.15)
### Added
- AMPL/Gurobi option for calculation Frobenius number based on Greenberg's algorithm (much faster than default algorithm).
- Apéry set calculation (requires AMPL and Gurobi).
- Command options for all features (see full list with --help).
- Generators of numerical semigroup now taken from file.

### Changed
- Old console options are no longer supported, use new command options.

### Removed
- Deleted /ampl directory. AMPL files are now generated when needed.

## 0.2.0 (2018.07.25)
### Added
- Calculation for the Sylvester denumerant.
- Calculation of the Sylvester denumerant up to a certain bound.

## 0.1.0 (2018.05.30)
### Added
- Preliminary version with files from older project.
- Frobenius number (slow algorithm).
- Set of gaps (slow algorithm).
- Membership problem (slow algorithm).


src/InputParser.h
//! @brief Parse C++ options
//! @author Diego Rodriguez Martinez (diego.rodriguez@usc.es)
//! Iain contributions from Stack Overflow
//! https://stackoverflow.com/questions/865668
//! http://iainhull.github.io


#include<vector>
#include<string>


#define NUM_REQUIRED_ARGS 1

class InputParser{

  struct Token {
    std::string str;
    bool opt;
  };

public:

    InputParser (int &argc, char **argv){

      if (argc > 1) {
        // Is there any argument?

        int start_point;

        // Positional arguments
        std::string str_st = std::string(argv[1]);
        if (argc > NUM_REQUIRED_ARGS and str_st.find_first_of("-") != 0) {
          this->problem = str_st;
          start_point = 2;
        }
        else {
          start_point = 1;
        }

        // Optional arguments
        for (int i=start_point; i < argc; ++i) {
          std::string str = std::string(argv[i]);
          if (str.find_first_of("-") == 0 and str.find_first_of("=") != std::string::npos) {
            // Si empieza por '-' y tiene un '=' en algún lado => descomponemos
            std::size_t pos_equal = str.find("=");
            std::string opt = str.substr (0,pos_equal);
            Token option_name;
            option_name.str = opt;
            option_name.opt = true;
            this->tokens.push_back(option_name);
            Token option_value;
            option_value.str = str.substr (pos_equal+1);
            option_value.opt = false;
            this->tokens.push_back(option_value);
          }
          else {
            Token option_name;
            option_name.str = str;
            if (str.find_first_of("-") == 0) {
              option_name.opt = true;
            } else {
              option_name.opt = false;
            }
            this->tokens.push_back(option_name);
          }
        }

        if (false) {
          std::cout << "ARGS:" << std::endl;
          for (unsigned i=0; i<this->tokens.size(); ++i) {
            std::cout << " + " << this->tokens[i].str;
            if (this->tokens[i].opt) std::cout << "*";
            std::cout << std::endl;
          }
        }
      }
    }

    bool isErrorCmd() const{
      for (unsigned i=0; i<this->tokens.size(); ++i) {
        if (this->tokens[i].str.find_first_of("=") == 0) {
          return true;
        }
      }
      return false;
    }

    const std::string getCmdArg(const std::string &option) const{

      unsigned i=0;
      while (i<this->tokens.size()) {
        if (this->tokens[i].opt) { // Solo mirarmos los que son opciones
          if (this->tokens[i].str.compare(option) == 0) {
            i++;
            if (i == this->tokens.size()) return ""; // Estamos al final
            if (this->tokens[i].opt) return ""; // No hay argumento
            return this->tokens[i].str;
          }
        }
        i++;
      }
      return "";
    }
    /// @author diego
    const std::string UnknownOptionFound(std::vector<std::string> list_of_knowns) const{
      std::string list_of_unknowns = "";
      for (unsigned i=0; i<this->tokens.size(); ++i) {
        if (this->tokens[i].opt) { // Solo miramos los que son opciones
          std::vector<std::string>::const_iterator itr;
          itr = std::find(list_of_knowns.begin(), list_of_knowns.end(), this->tokens[i].str);
          if (itr >= list_of_knowns.end()) {
            list_of_unknowns.append(this->tokens[i].str+" "); // Si no está en la lista de conocidos
          }
        }
      }
      return list_of_unknowns;
    }
    /// @author iain
    bool cmdOptionExists(const std::string &incumbent) const{
      for (unsigned i=0; i<this->tokens.size(); ++i) {
        if (this->tokens[i].opt) { // Solo miramos los que son opciones
          if (this->tokens[i].str.compare(incumbent) == 0) {
            return true;
          }
        }
      }
      return false;
    }

    bool cmdNoOptions() const {
      return this->tokens.empty();
    }

    bool cmdNoProblem() const {
      return this->problem.empty();
    }

    unsigned getOptionsSize() const {
      return this->tokens.size();
    }

    std::string getProblem() const {
      return this->problem;
    }

private:
    std::vector <Token> tokens;
    std::string problem = "";
    bool errors = false;
};


src/NumSemTests.h
/*!
 * @file    NumSemTests.h
 * @brief   NumSemTests class (definitions)
 * @author  Joaquin Ossorio-Castillo (joaquin.ossorio@usc.es)
 * @date    2018-06-01
 *
 */

#pragma once
#ifndef NUM_SEM_TESTS_H
#define NUM_SEM_TESTS_H

#include <fstream>
#include <sstream>

#include "NumericalSemigroup.h"

class NumSemTests
{
private:
    NumericalSemigroup* ns;
    int number_of_options;
public:

    //! @brief Constructor
    NumSemTests();

	//! @brief Empty destructor
	~NumSemTests();

    int calculate_sylvester_denumerant();

    NumericalSemigroup* initialize_numerical_semigroup_input();

    void numsem_init();

    void numsem_frobenius_number();

    void numsem_gaps();

    void numsem_membership();

    void numsem_sylvester_denumerant(int t);

    void print_numsem_options();

    void sylvester_polynomial_graph();

    void sylvester_denumerant_iter();

};

#endif


src/NumSemTests.cpp
/*!
 * @file    NumericalSemigroup.cpp
 * @brief   NumericalSemigroup class
 * @author  Joaquin Ossorio-Castillo (joaquin.ossorio@usc.es)
 * @date    2018-05-28
 *
 */

#include <chrono>

#include "NumSemTests.h"

NumSemTests::NumSemTests(){
    this->number_of_options = 7;
    this->numsem_init();
}

int NumSemTests::calculate_sylvester_denumerant() {
    /* calculate denumerant for t = */
    std::cout << "Introduce an integer for the Sylvester denumerant: ";
    int t;
    std::cin >> t;
    int den = this->ns->sylvester_denumerant(t, true);
    std::cout << std::endl;
    return den;
}

NumericalSemigroup* NumSemTests::initialize_numerical_semigroup_input() {

    std::cout << "### Add generators to the semigroup (type -1 for no more generators)" << std::endl;
    int newGenerator = 1;
    std::set<int> generators;
    while (newGenerator >= 0){
    	std::cout << "Generator " << generators.size() + 1 << ": ";
    	std::cin >> newGenerator;
    	std::cout << std::endl;
    	if (newGenerator > 1){
    		generators.insert(newGenerator);
    	} else if (newGenerator == 0 || newGenerator == 1) {
    		std::cout << "Generators must be different to 0 and 1 - Try again" << std::endl;
    	}
    }

    //std::cout << "NumericalSemigroup::NumericalSemigroup(std::set<int> generators)" << std::endl;
	NumericalSemigroup* ns = new NumericalSemigroup(generators);
    return ns;
}

void NumSemTests::numsem_init() {
    this->ns = this->initialize_numerical_semigroup_input();
    this->print_numsem_options();
    int option;
    std::cin >> option;
    std::cout << std::endl;
    while (option != this->number_of_options) {
        if (option > this->number_of_options || option < 1) {
            std::cout << "Please, select a valid option." << std::endl;
            std::cin >> option;
            std::cout << std::endl;
        } else if (option == 1) {
            this->numsem_frobenius_number();
            this->print_numsem_options();
            std::cin >> option;
            std::cout << std::endl;
        } else if (option == 2) {
            this->numsem_gaps();
            this->print_numsem_options();
            std::cin >> option;
            std::cout << std::endl;
        } else if (option == 3) {
            this->numsem_membership();
            this->print_numsem_options();
            std::cin >> option;
            std::cout << std::endl;
        } else if (option == 4) {
            this->numsem_sylvester_denumerant(1);
            this->print_numsem_options();
            std::cin >> option;
            std::cout << std::endl;
        } else if (option == 5) {
            this->sylvester_polynomial_graph();
            this->print_numsem_options();
            std::cin >> option;
            std::cout << std::endl;
        } else if (option == 6) {
            this->ns = this->initialize_numerical_semigroup_input();
            this->print_numsem_options();
            std::cin >> option;
            std::cout << std::endl;
        }
    }
}

void NumSemTests::numsem_frobenius_number() {
    std::cout << "f(S) = " << this->ns->frobenius_number() << std::endl;
}

void NumSemTests::numsem_gaps() {
    std::vector<int> set_gaps = this->ns->gaps();
    if (set_gaps.size() == 0) {
        std::cout << "G(S) = Ø" << std::endl;
    } else {
        std::cout << "G(S) = {" << set_gaps[0];
        for (int i = 1; i < set_gaps.size(); ++i) {
            std::cout << ", " << set_gaps[i];
        }
        std::cout << "}" << std::endl;
    }
}

void NumSemTests::numsem_membership() {
    std::cout << "Membership problem input: ";
    int memb_input;
    std::cin >> memb_input;
    std::cout << std::endl;
    if (this->ns->membership(memb_input)) {
        std::cout << "YES";
    } else {
        std::cout << "NO";
    }
    std::cout << std::endl;
}

void NumSemTests::numsem_sylvester_denumerant(int t) {
    int sd = this->ns->sylvester_denumerant(t,true);
    std::cout << "d(" << t << "; ";
    this->ns->print_generators();
    std::cout << ") = " << sd << std::endl << std::endl;
}

void NumSemTests::print_numsem_options() {
    std::cout << std::endl;
    this->ns->print_numerical_semigroup();
    std::cout << std::endl << std::endl << "Select an option:" << std::endl;
    std::cout << "\t" << "1. Calculate the Frobenius number of S." << std::endl;
    std::cout << "\t" << "2. Calculate the set of gaps of S." << std::endl;
    std::cout << "\t" << "3. Solve the membership problem for S and t." << std::endl;
    std::cout << "\t" << "4. Calculate the Sylvester denumerant d(t;S)." << std::endl;
    std::cout << "\t" << "5. Print to file the Sylvester denumerant graph d(t;S)." << std::endl;
    std::cout << "\t" << "6. Define a new numerical semigroup." << std::endl;
    std::cout << "\t" << "7. Exit." << std::endl;
}

void NumSemTests::sylvester_polynomial_graph() {

    std::ofstream ofs;
    ofs.open ("sylvester_polynomial.txt", std::ofstream::out | std::ofstream::trunc);
    ofs.close();

    int upper_bound;
    std::cout << "Upper bound for the Sylvester polynomial graph: ";
    std::cin >> upper_bound;
    std::cout << std::endl;

    for (int i = 1; i <= upper_bound; ++i){

        /* calculate denumerant */
        auto begin = std::chrono::high_resolution_clock::now();
        int denumerant = this->ns->sylvester_denumerant(i, false);
        auto end = std::chrono::high_resolution_clock::now();

        ofs.open ("sylvester_polynomial.txt", std::ofstream::out | std::ofstream::app);
        ofs << i << " " << denumerant << " " << this->ns->number_of_bits(i) << " " << this->ns->number_of_iterations() << " " << std::chrono::duration_cast<std::chrono::nanoseconds>(end-begin).count() << std::endl;
        ofs.close();
    }

}

void sylvester_denumerant_iter() {

    std::ofstream ofs;
    ofs.open ("sylvester_iters.txt", std::ofstream::out | std::ofstream::trunc);
    std::stringstream ss1;
    std::stringstream ss2;
    ss1 << "[";
    ss2 << "[";
}


src/NumericalSemigroup.h
/*!
 * @file    NumericalSemigroup.h
 * @brief   NumericalSemigroup class (definitions)
 * @author  Joaquin Ossorio-Castillo (joaquin.ossorio@usc.es)
 * @date    2018-05-28
 *
 */

#pragma once
#ifndef NUMERICAL_SEMIGROUP_H
#define NUMERICAL_SEMIGROUP_H

#include <iostream>
#include <set>
#include <vector>

class NumericalSemigroup
{
private:

	std::set<int> generators;	//!< System of generators of the numerical semigroup (default: NULL)
	bool is_num_sem;			//!< True if 'generators' is a valid system of generators (i.e., if gcd(a_1,...,a_n) = 1)
	int n_iters;

public:

    //! @brief Constructor
    NumericalSemigroup(std::set<int> generators);

	//! @brief Empty destructor
	~NumericalSemigroup();

	//! @brief  Adds a generator to the system of generators of the current numerical semigroup
	//! @param	generator  New generator
    //! @return True if the element was already inside the semigroup
	bool add_generator(int generator);

	//! @return Calculates the Frobenius number of the numerical semigroup
	int frobenius_number();
	int frobenius_number_ampl();

	//! @return Calculates the best lower bound for the Frobenius number from a number of criteria
	int frobenius_number_bound();

	//! @return The set of gaps of the numerical semigroup
	std::vector<int> gaps();

	//! @brief 	Calculates the greatest common divisor of two integers
	//! @param 	a First integers
	//! @param 	b Second integers
	//! @return gcd(a,b)
	int gcd(int a, int b);

	//! @return The number of generators of the current numerical semigroup
    int get_number_generators();

	//! @return The upper bounds for the ikp solutions of $t$ with respect to $a_1,...,a_n$
	std::vector<int> ikp_bounds(int t, std::vector<int> new_generators);

	//! @param 	lambda A vector of integers with the same size as the number of generators of this numerical semigroup
	//! @return The product $\sum_{i=1}^{n} \lambda_i a_i$
	int ikp_solution(std::vector<int> lambda, std::vector<int> new_generators);

	//! @return A bool that indicates if the present system of generators generates a numerical semigroup
	bool is_numerical_semigroup();

	//! @return True if t is inside the numerical NumericalSemigroup
	bool membership(int t);

	bool membership_core(int t, std::vector<int> new_generators);

	//! @return Next lambda iteration (used in the function sylvester_denumerant)
	std::vector<int> next_lambda(std::vector<int> lambda, std::vector<int> bounds);

	//! @brief Prints the system of generators
	void print_generators();

	//! @brief Prints the product $\sum_{i=1}^{n} \lambda_i a_i$ (not the result, which is given by ikp_solution)
	void print_ikp_solution(std::vector<int> lambda, std::vector<int> new_generators);

	//! @brief Prints the numerical semigroup in the format $S = <a_1,...,a_n>$
	void print_numerical_semigroup();

	//! @return The Sylvester denumerant of the integer $t$ with respect to the system of generators of the present numerical semigroup
	int sylvester_denumerant(int t, bool with_solutions);

	int number_of_bits(int t);

	int number_of_iterations();

	void write_ampl_numerical_semigroup_dat();

	void write_ampl_apery_set_member_mod();

	void write_ampl_apery_set_member_run(int s, int i);

	void write_ampl_apery_set_run(int s);

	void write_ampl_frobenius_number_run();

};

#endif


src/main.cpp
/*!
 * @file    main.cpp
 * @brief   main file
 * @author  Joaquin Ossorio-Castillo (joaquin.ossorio@usc.es)
 * @date    2018-05-28
 *
 */

#include <algorithm>
#include <cmath>
#include <iostream>
#include <fstream>
#include <math.h>

#include "InputParser.h"
#include "NumSemTests.h"
#include "NumericalSemigroup.h"

// int gcd (int a, int b);
// int apery_ilp (std::vector<int> sg_gens, int s, int i);
// int apery_zolp (std::vector<int> sg_gens, int s, int i);
// int apery_qubo (std::vector<int> sg_gens, int s, int i);

int main ( int argc, char *argv[] )
{

	std::cout << std::endl << "===========================================================================================" << std::endl;
	std::cout << " NumSem v0.3.0 (2018.10.15)" << std::endl;
	std::cout << " Copyright (C) Joaquín Ossorio-Castillo. All rights reserved." << std::endl;
	std::cout << " This software is free for non-commercial purposes." << std::endl;
	std::cout << " Full license information can be found in the LICENSE.txt file." << std::endl;
	std::cout << " https://github.com/jqnoc/numsem/" << std::endl;
	std::cout << "===========================================================================================" << std::endl;

	system("rm -rf .tmp; mkdir .tmp");

	/* define options */
	std::vector<std::string> list_of_knowns;

	list_of_knowns.push_back("--help");   // help
    bool opt_help = false;

	std::string opt_file;

	list_of_knowns.push_back("-f");  // Frobenius problem
	bool opt_frobenius;

	list_of_knowns.push_back("-g");  // set of gaps
	bool opt_gaps;

	list_of_knowns.push_back("-d");  // Sylvester denumerant
	list_of_knowns.push_back("-ds");  // Sylvester denumerant (with solutions)
	int opt_sylvester_denumerant;

	list_of_knowns.push_back("-dg");  // Sylvester denumerant graph
	int opt_sylvester_denumerant_graph;

	list_of_knowns.push_back("-m");  // numerical semigroup membership
	int opt_nsmp;

	list_of_knowns.push_back("-ampl");  // use AMPL
	bool opt_ampl = false;

	/* parse arguments */
	try {
		InputParser input(argc, argv);

		/* check for empty options */
		if (input.cmdNoProblem() and input.cmdNoOptions()) {
            throw "ERROR: no arguments";
        }

		/* check for errors in options */
		if (input.isErrorCmd()) {
            throw "ERROR: wrong command line syntax";
        }

		/* check for help option */
		if (input.cmdOptionExists("--help")) {
            if (not input.getCmdArg("--help").empty()){
                throw "ERROR: --help option cannot take arguments";
            }
            opt_help = true;
            if (input.getOptionsSize() > 1) {
                std::cout << "Option --help detected: other options are ignored" << std::endl;
            }
        throw "numsem help";
        }

		/* check for file */
		if (input.cmdNoProblem()) {
			throw "ERROR: wrong command line syntax (file must be first argument)";
		}

		/* initialize numerical semigroup from file */
		opt_file = input.getProblem();
		std::fstream ns_file(opt_file, std::ios_base::in);
		int new_generator;
		std::set<int> generators;
		while (ns_file >> new_generator) {
			if (new_generator > 1) {
				generators.insert(new_generator);
			}
		}
		NumericalSemigroup* ns = new NumericalSemigroup(generators);
		std::cout << "Numerical semigroup: ";
		ns->print_numerical_semigroup();
		std::cout << std::endl;
		std::cout << "Total number of generators: " << ns->get_number_generators() << std::endl;

		/* check for unknown options */
		std::string unknown_options = input.UnknownOptionFound(list_of_knowns);
        if (not unknown_options.empty()) {
			throw "ERROR: unknown options";
        }

		/* AMPL */
		if (input.cmdOptionExists("-ampl")) {
            if (not input.getCmdArg("-ampl").empty()) {
                throw "ERROR: option -ampl cannot take arguments";
            }
			opt_ampl = true;
        }

		/* Frobenius problem */
		if (input.cmdOptionExists("-f")) {
            if (not input.getCmdArg("-f").empty()) {
                throw "ERROR: option -f cannot take arguments";
            }
			opt_frobenius = true;
			std::cout << "f(S) = ";
			if (opt_ampl) {
				std::cout << ns->frobenius_number_ampl();
			} else {
				std::cout << ns->frobenius_number();
			}
			std::cout << std::endl;
        }

		/* set of gaps */
		if (input.cmdOptionExists("-g")) {
            if (not input.getCmdArg("-g").empty()){
                throw "ERROR: option -g cannot take arguments";
            }
			opt_gaps = true;
			std::vector<int> set_gaps = ns->gaps();
		    if (set_gaps.size() == 0) {
		        std::cout << "G(S) = Ø" << std::endl;
		    } else {
		        std::cout << "G(S) = {" << set_gaps[0];
		        for (int i = 1; i < set_gaps.size(); ++i) {
		            std::cout << ", " << set_gaps[i];
		        }
		        std::cout << "}" << std::endl;
		    }
        }

		/* Sylvester denumerant */
		if (input.cmdOptionExists("-d")) {
            std::string str_sylvester_denumerant = input.getCmdArg("-d");
            if (str_sylvester_denumerant.empty()){
                throw "ERROR: option -d needs an argument";
            }
			opt_sylvester_denumerant = stoi(str_sylvester_denumerant);
			if (opt_sylvester_denumerant < 1) {
				throw "ERROR: option -d needs an integer greater than 0";
			}
			std::cout << "d(" << opt_sylvester_denumerant << ") = " << ns->sylvester_denumerant(opt_sylvester_denumerant,false) << std::endl;
        }

		/* Sylvester denumerant (with solutions) */
		if (input.cmdOptionExists("-ds")) {
            std::string str_sylvester_denumerant = input.getCmdArg("-ds");
            if (str_sylvester_denumerant.empty()){
                throw "ERROR: option -ds needs an argument";
            }
			opt_sylvester_denumerant = stoi(str_sylvester_denumerant);
			if (opt_sylvester_denumerant < 1) {
				throw "ERROR: option -d needs an integer greater than 0";
			}
			std::cout << "d(" << opt_sylvester_denumerant << ") = " << ns->sylvester_denumerant(opt_sylvester_denumerant,true) << std::endl;
        }
	} catch (char const *e) {
		if (not std::string(e).empty()) {
			std::cout << std::endl;
            std::cout << e << std::endl;
            std::cout << " Use:" << std::endl;
            std::cout << "   numsem *file* [options]" << std::endl << std::endl;
            std::cout << " Available options:" << std::endl;
            std::cout << "   --help: print the help" << std::endl;
            std::cout << "   *file*: the file with the generators of the numerical semigroup S" << std::endl;
            std::cout << "   -f: calculate Frobenius number of S" << std::endl;
            std::cout << "   -g: calculate set of gaps of S" << std::endl;
            std::cout << "   -d *t*: calculate Sylvester denumerant for t and S" << std::endl;
            std::cout << "   -ds *t*: calculate Sylvester denumerant for t and S and print all solutions" << std::endl;
            std::cout << "   -dg *b*: calculate Sylvester function from 0 to b" << std::endl;
            std::cout << "   -m *t*: calculate if t is in S" << std::endl;
            std::cout << std::endl;
			exit(0);
		}
	}

	/* read generators from file */

	//NumSemTests* nst = new NumSemTests();

	return 1;
}
